docs/extensibility/custom-component.md

```diff
@@ -0,0 +1,146 @@
+
+:::code source="snippets/MailDevResource/MailKit.Client/MailKitClientFactory.cs":::
+
+The `MailKitClientFactory` class is a factory that creates an `ISmtpClient` instance based on the configuration settings. It's responsible for returning an `ISmtpClient` implementation that's connected to a configured SMTP server and optionally authenticated. Next, you need to expose the functionality for the consumers to register this factory with the dependency injection container. Add the following code to the `MailKit.Client` project in a file named MailKitClientServiceCollectionExtensions.cs_:
+
+:::code source="snippets/MailDevResource/MailKit.Client/MailKitClientServiceCollectionExtensions.cs":::
+
+The preceding code adds two extension methods on the `IHostApplicationBuilder` type, one for the standard registration of MailKit and another for keyed-registration of MailKit. Both extensions ultimately rely on the private `AddMailKitClient` method to register the `MailKitClientFactory` with the dependency injection container as a scoped-life service. The reason for registering the `MailKitClientFactory` as a scoped service is due to the fact that the connection (and authentication) operations are considered expensive and should be reused within the same scope where possible. In other words, for a single request, the same `ISmtpClient` instance should be used.
+
+The pattern for binding component settings to the builder's configuration, is to first instantiate the settings class and then bind settings to the specific section of the configuration.
+
+After they're bound to the existing configuration, the optional `configureSettings` delegate is invoked with the `settings` instance. This allows the consumer to further configure the settings, ensuring that manual code settings are honored over configuration settings. Next, considering whether or not the `serviceKey` value was provided, the `MailKitClientFactory` is registered with the dependency injection container as either a standard or keyed service.
+
+> [!IMPORTANT]
+> It's intentional that the `implementationFactory` overload is called when registering services. The `CreateMailKitClientFactory` method throws when the configuration is invalid. This ensures that creation of the `MailKitClientFactory` is deferred until it's needed and it prevents the app from erroring out before logging is available.
+
+The registration of health checks, and telemetry are described in a bit more detail in the following sections.
+
+### Add health checks
```

Comment:
We may want to link to https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks to recommend reusing an existing health check library (if there is one).

Reply:
Good callout, I added a link to the Health Checks UI sample.